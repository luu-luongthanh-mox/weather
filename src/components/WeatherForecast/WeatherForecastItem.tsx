import React from "react";
import { Card, Col, Row } from "react-bootstrap";
import classnames from "classnames";

import { ForecastWeather } from "@models/ForecastWeather";
import { useDisplayTemp } from "@hooks/useDisplayTemp";
import { useWeatherIcon } from "@hooks/useWeatherIcon";
import { useGetDayName } from "@hooks/useGetDayName";
import { isToday } from "@utils/date";

import styles from "./weatherForecast.module.css";

interface WeatherForecastProps {
  data: ForecastWeather;
}

export const WeatherForecastItem: React.FC<WeatherForecastProps> = ({
  data,
}) => {
  const minTemp = useDisplayTemp(data.min_temp);
  const maxTemp = useDisplayTemp(data.max_temp);
  const iconSrc = useWeatherIcon(data.weather_state_abbr);
  const dayName = useGetDayName(data.applicable_date);
  const isActive = isToday(new Date(data.applicable_date));
  return (
    <Col xs={12} sm={6} md={4} lg>
      <Card>
        <Card.Body className={styles.itemContent}>
          <Card.Title className={isActive ? styles.today : ""}>
            {dayName}
          </Card.Title>
          <Card.Img src={iconSrc} width="50" className={styles.icon} />
          <Row>
            <Col className={classnames(styles.temp, styles.tempHight)}>
              {maxTemp}
            </Col>
            <Col className={styles.temp}>{minTemp}</Col>
          </Row>
        </Card.Body>
      </Card>
    </Col>
  );
};
