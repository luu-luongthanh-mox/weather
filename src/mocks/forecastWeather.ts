import { ForecastWeather } from "@models/ForecastWeather";

export const forecastWeatherItemMock: ForecastWeather = {
  id: 4964038678151168,
  weather_state_name: "Light Cloud",
  weather_state_abbr: "c",
  wind_direction_compass: "NNE",
  created: "2020-12-06T00:31:57.873155Z",
  applicable_date: "2020-12-04",
  min_temp: 15.719999999999999,
  max_temp: 21.325,
  the_temp: 20.03,
  wind_speed: 6.226959035706144,
  wind_direction: 31.834062016955453,
  air_pressure: 1021.5,
  humidity: 53,
  visibility: 13.535328680505845,
  predictability: 70,
};
