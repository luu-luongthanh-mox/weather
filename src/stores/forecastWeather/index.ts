export { forecastWeatherSlice, getForecastWeatherById } from "./reducer";
