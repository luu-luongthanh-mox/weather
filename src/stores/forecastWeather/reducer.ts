import axios from "axios";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

import { WeatherLocation } from "@models/ForecastWeather";
import { getWeatherForecastById } from "@services/MetaWeather";
import { StateStatus } from "@models/State";

interface State {
  data: WeatherLocation;
  loading: string;
  currentRequestId: string | undefined;
  error?: any;
}

const initialState: State = {
  data: {
    consolidated_weather: [],
    latt_long: "",
    location_type: "",
    sun_rise: "",
    sun_set: "",
    time: "",
    timezone: "",
    timezone_name: "",
    title: "",
    woeid: 0,
  },
  loading: StateStatus.idle,
  currentRequestId: undefined,
  error: undefined,
};
const GET_BY_ID_ACTION = "forecastWeather/getById";
export const getForecastWeatherById = createAsyncThunk<any, string>(
  GET_BY_ID_ACTION,
  async (id, { signal }) => {
    if (id) {
      const source = axios.CancelToken.source();
      signal.addEventListener("abort", () => {
        source.cancel();
      });
      const response = await getWeatherForecastById(id, {
        cancelToken: source.token,
      });
      return response;
    }
  }
);

// Ref: https://redux-toolkit.js.org/tutorials/basic-tutorial
// Slice can create actions and reducer together
export const forecastWeatherSlice = createSlice({
  name: "forecastWeather",
  initialState: initialState,
  reducers: {
    storeIdToken: (state, action: PayloadAction<string>) => {
      return { ...state, idToken: action.payload };
    },
  },
  extraReducers: {
    [GET_BY_ID_ACTION + "/pending"]: (state, action) => {
      if (state.loading === StateStatus.idle) {
        state.loading = StateStatus.loading;
        state.currentRequestId = action.meta.requestId;
      }
    },
    [GET_BY_ID_ACTION + "/fulfilled"]: (state, action) => {
      const { requestId } = action.meta;
      if (
        state.loading === StateStatus.loading &&
        state.currentRequestId === requestId
      ) {
        state.loading = StateStatus.idle;
        state.data = action.payload;
        state.currentRequestId = undefined;
      }
    },
    [GET_BY_ID_ACTION + "/rejected"]: (state, action) => {
      const { requestId } = action.meta;
      if (
        state.loading === StateStatus.loading &&
        state.currentRequestId === requestId
      ) {
        state.loading = StateStatus.idle;
        state.error = action.error;
        state.currentRequestId = undefined;
      }
    },
  },
});
