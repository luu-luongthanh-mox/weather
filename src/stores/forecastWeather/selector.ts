import { RootState } from "@stores";

export const getForecastWeather = (state: RootState) => state.forecastWeather;

export const getForecastWeatherData = (state: RootState) =>
  state.forecastWeather.data?.consolidated_weather || [];
