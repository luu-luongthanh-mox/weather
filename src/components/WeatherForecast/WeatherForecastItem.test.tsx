import React from "react";
import { render, screen } from "@testing-library/react";

import { forecastWeatherItemMock } from "@mocks/forecastWeather";

import { WeatherForecastItem } from "./WeatherForecastItem";

describe("Component: WeatherForecastItem", () => {
  it("renders the data", () => {
    const { asFragment } = render(
      <WeatherForecastItem data={forecastWeatherItemMock} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders max temp", () => {
    render(<WeatherForecastItem data={forecastWeatherItemMock} />);
    const linkElement = screen.getByText(/21/i);
    expect(linkElement).toBeInTheDocument();
  });

  it("renders min temp", () => {
    render(<WeatherForecastItem data={forecastWeatherItemMock} />);
    const linkElement = screen.getByText(/16/i);
    expect(linkElement).toBeInTheDocument();
  });

  it("renders day name", () => {
    render(<WeatherForecastItem data={forecastWeatherItemMock} />);
    const linkElement = screen.getByText(/Fri/i);
    expect(linkElement).toBeInTheDocument();
  });
});
