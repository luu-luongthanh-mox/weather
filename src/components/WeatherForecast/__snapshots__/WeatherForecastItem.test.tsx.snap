// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: WeatherForecastItem renders the data 1`] = `
<DocumentFragment>
  <div
    class="col-lg col-md-4 col-sm-6 col-12"
  >
    <div
      class="card"
    >
      <div
        class="itemContent card-body"
      >
        <div
          class="card-title h5"
        >
          Fri
        </div>
        <img
          class="card-img icon"
          src="/weather/c.svg"
          width="50"
        />
        <div
          class="row"
        >
          <div
            class="temp tempHight col"
          >
            21°
          </div>
          <div
            class="temp col"
          >
            16°
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
